### PR TITLE
OSX: implement sensors_battery

### PR DESCRIPTION
**Pull request by wiggin15**

### CODE REVIEW

File: psutil/_psutil_osx.c [hunk 1]
```diff
@@ -1788,6 +1790,93 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
...
+    if (!power_info) {
+        PyErr_SetString(PyExc_RuntimeError,
+            "Failed to get power sources info");
+        goto error;
+    }
+
+    power_sources_list = IOPSCopyPowerSourcesList(power_info);
+    if (!power_sources_list) {
+        PyErr_SetString(PyExc_RuntimeError,
+            "Failed to get power sources list");
```
Comment (giampaolo) [inline psutil/_psutil_osx.c]: I would change the err message to "IOPSCopyPowerSourcesList() syscall failed"

File: psutil/_psutil_osx.c [hunk 2]
```diff
@@ -1788,6 +1790,93 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
...
+    if (!CFNumberGetValue(time_to_empty_ref,
+                          kCFNumberIntType, &time_to_empty)) {
+        /* This value is recommended for non-Apple power sources, so it's not
+         * an error if it doesn't exist. We'll return -1 for "unknown" */
+        /* A value of -1 indicates "Still Calculating the Time" also for
+         * apple power source */
+        time_to_empty = -1;
+    }
+
+    py_retlist = Py_BuildValue("Iii",
```
Comment (giampaolo) [inline psutil/_psutil_osx.c]: I think ` Py_BuildValue("Iii"` produces a tuple, not a list.

File: psutil/_psutil_osx.c [hunk 3]
```diff
@@ -1788,6 +1790,93 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
...
+    if (!power_sources_list) {
+        PyErr_SetString(PyExc_RuntimeError,
+            "Failed to get power sources list");
+        goto error;
+    }
+
+    /* Should only get one source. But in practice, check for > 0 sources */
+    if (!CFArrayGetCount(power_sources_list)) {
+        PyErr_SetString(PyExc_RuntimeError,
+            "Power sources list is empty");
```
Comment (giampaolo) [inline psutil/_psutil_osx.c]: I checkout out your branch and bumped into this error on my virtualized OSX box.
It seems this happens where there is not battery so we probably have to return None here.

**Review by giampaolo — CHANGES_REQUESTED** [review]
Cool stuff. Some comments inline. Also another thing. In the future we'll probably want to make `sensors_battery` return info about multiple batteries as a list. Are these APIs able to do that?

Also, it would be good to have a specific test comparing results with a CLI tool. See here:
http://osxdaily.com/2015/12/10/get-mac-battery-life-info-command-line-os-x/

File: psutil/_psutil_osx.c (hunk 2 above)
Comment (wiggin15) [reply psutil/_psutil_osx.c]: I'm following the style of the rest of psutil. the name `py_retlist` is used for tuples as well as lists everywhere in the code.
There is no variable currently named "py_rettuple" or "py_retobj" etc. What name would you suggest?

Comment (giampaolo) [reply psutil/_psutil_osx.c]: > I'm following the style of the rest of psutil. the name py_retlist is used for tuples as well as lists everywhere in the code.

Really? I didn't realize that. Anyway, I suggest `py_tuple` in this case.

File: psutil/_psutil_osx.c [hunk 4]
```diff
@@ -1788,6 +1790,93 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
...
+    CFStringRef ps_state_ref = NULL;
+    uint32_t capacity;     /* units are percent */
+    int time_to_empty;     /* units are minutes */
+    int is_power_plugged;
+
+    power_info = IOPSCopyPowerSourcesInfo();
+
+    if (!power_info) {
+        PyErr_SetString(PyExc_RuntimeError,
+            "Failed to get power sources info");
```
[the remark this answers is not in the document]
Comment (wiggin15) [reply psutil/_psutil_osx.c]: I was following the style of "psutil_disk_io_counters" where errors for "IO*" functions set strings like "unable to get the disk properties" - but I will change this.

File: psutil/_psutil_osx.c (hunk 3 above)
Comment (wiggin15) [reply psutil/_psutil_osx.c]: Already done :)

**Comment by wiggin15** [comment]
> Are these APIs able to do that?

Since the API returns "PowerSourcesList", I assume this can include more than one source, but I will have no way to test this.

**Comment by giampaolo** [comment]
> Since the API returns "PowerSourcesList", I assume this can include more than one source, but I will have no way to test this.

OK, let's ignore this for now. But just FYI my plan is to have:

`sensors_battery(all=True) -> [ntuple(...), ...]`

File: psutil/_psutil_osx.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (giampaolo) [reply psutil/_psutil_osx.c]: OK, I will change those ones then. In general I will probably want to apply this paradigm everywhere.

**Comment by wiggin15** [comment]
Updated commit

File: psutil/_psosx.py [hunk 5]
```diff
@@ -207,6 +207,29 @@ def disk_partitions(all=False):
...
+
+
+def sensors_battery():
+    """Return battery information.
+    """
+    try:
+        percent, minsleft, power_plugged = cext.sensors_battery()
+    except NotImplementedError:
+        # no power source - return None according to interface
+        return None
```
Comment (giampaolo) [inline psutil/_psosx.py]: nice solution, I like it

File: psutil/tests/__init__.py [hunk 6]
```diff
@@ -70,7 +70,7 @@
     'VERBOSITY',
     "HAS_CPU_AFFINITY", "HAS_CPU_FREQ", "HAS_ENVIRON", "HAS_PROC_IO_COUNTERS",
     "HAS_IONICE", "HAS_MEMORY_MAPS", "HAS_PROC_CPU_NUM", "HAS_RLIMIT",
-    "HAS_SENSORS_BATTERY", "HAS_BATTERY""HAS_SENSORS_FANS",
+    "HAS_SENSORS_BATTERY", "HAS_BATTERY", "HAS_SENSORS_FANS",
```
Comment (giampaolo) [inline psutil/tests/__init__.py]: good catch

File: psutil/tests/test_osx.py [hunk 7]
```diff
@@ -285,6 +286,18 @@ def test_net_if_stats(self):
...
+
+    @unittest.skipIf(not HAS_BATTERY, "no battery")
+    def test_sensors_battery(self):
+        out = sh("pmset -g batt")
+        percent = re.search("(\d+)%", out).group(1)
+        drawing_from = re.search("Now drawing from '([^']+)'", out).group(1)
+        power_plugged = drawing_from == "AC Power"
+        psutil_result = psutil.sensors_battery()
+        self.assertEqual(psutil_result.power_plugged, power_plugged)
+        self.assertEqual(psutil_result.percent, int(percent))
```
Comment (giampaolo) [inline psutil/tests/test_osx.py]: sweet!

**Pull request merged**